GH+CCE for a gauge wave

Sizheng-Ma opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

This PR adds a GH+CCE executable for a gauge wave (without horizons). It lays the foundation for CCM. The system is **not** compatible with the local time stepping.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
I add one more template to `SendGhWorldtubeData` so that it can distinguish between `LTS` and `GTS`. Please remember to update your `CceWorldtubeTarget` @knelli2 
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

Sizheng-Ma committed:
Add GH+CCE executable for a gauge wave

Sizheng-Ma committed:
Reorder the phase list

knelli2 committed:
Send GH data to CCE Action

knelli2 committed:
Change callback to avoid interface manager

Sizheng-Ma committed:
Make CharacteristicExtractDefaults compatible with GH+CCE

knelli2 commented on src/ParallelAlgorithms/Interpolation/Callbacks/SendGhWorldtubeData.hpp:
```diff
@@ -32,7 +33,7 @@ namespace callbacks {
 ///
 /// \note This callback requires the temporal ID in an InterpolationTargetTag be
 /// a TimeStepId.
-template <typename CceEvolutionComponent, bool DuringSelfStart>
+template <typename CceEvolutionComponent, bool DuringSelfStart, bool LTS>
```
I see why you had to add the extra template, but `DuringSelfStart` and `LTS` serve the same purpose here so could you combine them somehow? Maybe name it `NeedsInterfaceManager`? Then whatever calls this is responsible for passing the correct bool.

Sizheng-Ma replied:
I'd think both `DuringSelfStart` and `LTS` are wanted because we need to distinguish self-start from evolution even when the interface manager is used (`Cce::Actions::ReceiveGhWorldtubeData` takes `DuringSelfStart` other than `NeedsInterfaceManager`).

knelli2 replied:
Ahh I see, then I think this `LTS` template is ok.

knelli2 commented on src/Evolution/Executables/GeneralizedHarmonic/GeneralizedHarmonicBase.hpp:
```diff
@@ -145,19 +145,21 @@ namespace detail {
 template <typename InitialData>
 constexpr auto make_default_phase_order() {
```
Capitalize `reorder` in the commit message please.

knelli2 commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonicCce.hpp:
```diff
@@ -0,0 +1,273 @@
+}  // namespace Frame
+namespace Parallel {
+template <typename Metavariables>
+class CProxy_ConstGlobalCache;
```
What is this? We don't have a `ConstGlobalCache`

knelli2 commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonicCce.hpp:
```diff
@@ -0,0 +1,273 @@
+              tmpl::list<Actions::RunEventsAndTriggers, Actions::ChangeSlabSize,
+                         step_actions<false>, Actions::AdvanceTime,
+                         PhaseControl::Actions::ExecutePhaseChange>>>>>;
+
+  template <bool DuringSelfStart, bool LTS>
```
If you make the above change about the `LTS` template to `SendGhWorldtubeData`, I don't think you need this `LTS` template. Just use the `local_time_stepping` member that's defined at the top of the metavariables.

For the `SendGhWorldtubeData` template, I think you can use `tmpl::or_v<DuringSelfStart, local_time_stepping>`.

knelli2 replied:
I still think you can get rid of this `LTS` template. Just use the `local_time_stepping` bool directly inside the target.

Sizheng-Ma replied:
Done!

knelli2 commented on tests/InputFiles/GeneralizedHarmonic/GaugeWave3DCce.yaml:
```diff
@@ -0,0 +1,200 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+# Executable: EvolveGhCceGaugeWave
+# Check: parse;execute
+# Timeout: 8
```
I'd remove the `execute` part. This takes a while. Parsing should be ok for now

knelli2 reviewed (commented):
Ok all changes look good! Please squash and rebase.

Seems like GCC 9 is still running out of memory on CI...not much I think we can do about that at the moment. One idea could be, for the time being, to just not include this target when building the tests

knelli2 commented:
@Sizheng-Ma Once #4299 has been merged, you'll have to rebase and fix the `#include`s.

knelli2 commented:
@Sizheng-Ma This needs a rebase on develop. Several things have changed over the past month or so.

In order to have CI actually pass and not run out of memory trying to build this executable, please remove the input file you have. If there isn't an input file in the test directory, then it won't build the executable. Please make sure this executable builds on your own branch before pushing. I'll also check that it will build in a couple places before I approve. This will just be a temporary fix until we can reduce the memory usage during compile time.

Sizheng-Ma commented:
> @Sizheng-Ma This needs a rebase on develop. Several things have changed over the past month or so.
> 
> In order to have CI actually pass and not run out of memory trying to build this executable, please remove the input file you have. If there isn't an input file in the test directory, then it won't build the executable. Please make sure this executable builds on your own branch before pushing. I'll also check that it will build in a couple places before I approve. This will just be a temporary fix until we can reduce the memory usage during compile time.

Sizheng-Ma commented:
> @Sizheng-Ma This needs a rebase on develop. Several things have changed over the past month or so.
> 
> In order to have CI actually pass and not run out of memory trying to build this executable, please remove the input file you have. If there isn't an input file in the test directory, then it won't build the executable. Please make sure this executable builds on your own branch before pushing. I'll also check that it will build in a couple places before I approve. This will just be a temporary fix until we can reduce the memory usage during compile time.

I've updated the branch and removed the input file. @knelli2

knelli2 reviewed (dismissed):
This is still dependent on #4466, but I've built both of these executables locally on this branch and after rebasing on the current develop, so looks good!

nilsdeppe commented on src/Evolution/Systems/Cce/Callbacks/SendGhWorldtubeData.hpp:
```diff
@@ -41,7 +42,7 @@ namespace callbacks {
 /// a TimeStepId.
 template <typename CceEvolutionComponent, typename InterpolationTargetTag,
-          bool DuringSelfStart>
+          bool DuringSelfStart, bool LTS>
```
`LTS` -> `LocalTimeStepping`

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_SendGhWorldtubeData.cpp:
```diff
@@ -28,7 +29,7 @@
 namespace {
 
-template <typename InterpolationTargetTag>
+template <typename InterpolationTargetTag, bool LTS>
```
`LTS` -> `LocalTimeStepping`

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_SendGhWorldtubeData.cpp:
```diff
@@ -89,16 +90,21 @@ template <typename Metavariables>
       Parallel::PhaseActions<Parallel::Phase::Initialization, tmpl::list<>>>;
 };
 
+template <bool LTS>
```
`LTS` -> `LocalTimeStepping`

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_SendGhWorldtubeData.cpp:
```diff
@@ -108,27 +114,26 @@ struct test_metavariables {
-    "Unit.NumericalAlgorithms.Interpolation.SendGhWorldtubeDataCallback",
-    "[Unit][Cce]") {
-  MAKE_GENERATOR(gen);
+template <bool LTS, typename Generator>
```
`LTS` -> `LocalTimeStepping`

Sizheng-Ma replied:
Done! Cheers.

nilsdeppe reviewed (changes requested):
You can squash immediately. Also please rebase on the latest develop to make sure you have some recent changes to the GH system
